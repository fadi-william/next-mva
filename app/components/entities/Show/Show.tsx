// The imported libs.
import { observer } from "mobx-react";
import * as React from "react";

// The component's model type.
import { TShow } from "../../../models/show/Show";

// The Show's props interface.
interface IShowProps {
  show: TShow;
}

// Import the component styles.
import "./Show.scss";

@observer
class Show extends React.Component<IShowProps> {
  public render() {
    const { show } = this.props;
    const { isLoading } = show;

    return (
      <div className="Show">
        {!isLoading && (
          <div>
            <h1>{show.name}</h1>
            <p>{show.summary}</p>
            <img src={show.image} />
          </div>
        )}
        {isLoading && <div>Loading...</div>}
      </div>
    );
  }
}

export default Show;
