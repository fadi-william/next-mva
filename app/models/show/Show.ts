// The imported libs.
import { applySnapshot, flow, types } from "mobx-state-tree";
import { getShow } from "../../service/showService";

export type TShow = typeof Show.Type;

// The module's global variables.
let store: TShow = null;

// If the id changed, we will re-initialize the store *(on the client-side).
let currentShowId: number = null;

export const Show = types.model({
    id: types.maybe(types.identifier(types.number)),
    image: types.maybe(types.string),
    isLoading: types.optional(types.maybe(types.boolean), true),
    name: types.maybe(types.string),
    summary: types.maybe(types.string),
})
.actions( (self) => ({
    doneLoading() {
        self.isLoading = false;
    },
}))
.actions( (self) => ({
    getShow: flow(function *() {
        const response = yield getShow(self.id);

        if (response.data) {
            const show = Object.assign({}, response.data);

            show.id = Number(response.data.id);
            show.image = response.data.image.medium;

            applySnapshot(self, show);

            self.doneLoading();
        }
    }),
}))
.actions( (self) => ({
    afterCreate() {
        self.getShow();
    },
}));

// Utility function - Instantiates a part of the mobx-state-tree store.
export function getStore(isServer, showId, snapshot = null) {
    // If this is the server.
    if (isServer) {
      currentShowId = showId;
      store = Show.create({
          id: showId,
      });
    }
    if (store === null || currentShowId !== showId) {
        currentShowId = showId;
        store = Show.create({
            id: showId,
        });
    }
    if (snapshot) {
      applySnapshot(store, snapshot);
    }
    return store;
}
