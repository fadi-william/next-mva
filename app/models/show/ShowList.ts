// The imported libs.
import { applySnapshot, types  } from "mobx-state-tree";
import "rxjs/add/operator/debounceTime";
import "rxjs/add/operator/map";
import { Subject } from "rxjs/Subject";

import { getShows } from "../../service/showService";
import { Show } from "./Show";

export type TShowList = typeof ShowList.Type;

// The module's global variables.
let store: TShowList = null;
const getShowsSubject$ = new Subject<void>();

export const ShowList = types.model({
    isLoading: types.optional(types.maybe(types.boolean), true),
    shows: types.optional(types.array(Show), []),
})
// The shows model's related actions.
.actions( (self) => ({
    doneLoading() {
        self.isLoading = false;
    },
    getShows() {
        // Get the latest shows.
        getShowsSubject$.next();
    },
}))
.actions( (self) => ({
    afterCreate() {
        getShowsSubject$
            .debounceTime(200)
            .map(() => {
                return getShows();
            })
            .subscribe((showsService) => {
                showsService.then((showResponse) => {
                    const shows = showResponse.data.map( (showObj) => {
                        return {
                            id: showObj.show.id,
                            name: showObj.show.name,
                        };
                    });

                    applySnapshot(self.shows, shows);

                    if (self.isLoading) {
                        self.doneLoading();
                    }
                });
            });

        // Trigger get shows.
        self.getShows();
    },
    beforeDestroy() {
        getShowsSubject$.unsubscribe();
    },
}))
// Derived data from the shows model.
.views( (self) => ({
    get totalShows() {
        return self.shows.length;
    },
}));

// Utility function - Instantiates a part of the mobx-state-tree store.
export function getStore(isServer, snapshot = null) {
    // If this is the server.
    if (isServer) {
      store = ShowList.create();
    }
    if (store === null) {
        store = ShowList.create();
    }
    if (snapshot) {
      applySnapshot(store, snapshot);
    }
    return store;
}
