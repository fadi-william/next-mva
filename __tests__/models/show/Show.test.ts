import { getSnapshot } from "mobx-state-tree";
import { getStore, Show } from "../../../app/models/show/Show";

// Mock the show service.
jest.mock("../../../app/service/showService");

const show = {
    id: 975,
    image: "https://img.url/1",
    isLoading: true,
    name: "A show",
    summary: "This is the show summary",
};

const show2 = {
    id: 481,
    image: "https://img.url/2",
    isLoading: true,
    name: "A second show",
    summary: "This is the second show summary",
};

describe("show model tests", () => {

    it("can create a show model", () => {
        const showStore = Show.create(show);

        expect(showStore.id).toBe(975);
        expect(showStore.isLoading).toBe(true);
        expect(getSnapshot(showStore)).toMatchSnapshot();
    });

    it("can load the show from a web service", async () => {
        const showStore = Show.create({
            id: 975,
        });
        await showStore.getShow();

        expect(showStore.isLoading).toBe(false);
        expect(showStore.id).toBe(975);
    });

    it("can get a store on the client", () => {
        const showStore = getStore(false, 975);

        expect(showStore.id).toBe(975);
    });

    it("can get a store on the server", () => {
        const showStore = getStore(true, 975);

        expect(showStore.id).toBe(975);
    });

    it("can restore a show model from a snaphost", () => {

        const showStore = getStore(false, 975, show);

        expect(showStore.name).toBe("A show");
    });

    it("can re-instantiate the store when the show change", () => {
        const showStore = getStore(true, 975, show);
        expect(showStore.name).toBe("A show");

        const showStore2 = getStore(false, 481, show2);
        expect(showStore2.id).toBe(481);
        expect(showStore2.name).toBe("A second show");
    });
});
