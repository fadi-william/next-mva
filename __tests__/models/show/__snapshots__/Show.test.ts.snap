// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`show model tests can create a show model 1`] = `
Object {
  "id": 975,
  "image": "https://img.url/1",
  "isLoading": true,
  "name": "A show",
  "summary": "This is the show summary",
}
`;
